src/shogun/multiclass/tree/CARTree.cpp

```diff
@@ -69,97 +73,121 @@ CARTree::CARTree(SGVector<bool> attribute_types, EProblemType prob_type, int32_t
 
 CARTree::~CARTree()
 {
-
 }
 
 void CARTree::set_labels(std::shared_ptr<Labels> lab)
 {
-	if (lab->get_label_type()==LT_MULTICLASS)
+	if (lab->get_label_type() == LT_MULTICLASS)
 		set_machine_problem_type(PT_MULTICLASS);
-	else if (lab->get_label_type()==LT_REGRESSION)
+	else if (lab->get_label_type() == LT_REGRESSION)
 		set_machine_problem_type(PT_REGRESSION);
 	else
 		error("label type supplied is not supported");
 
-
-
-	m_labels=lab;
+	m_labels = lab;
 }
 
 void CARTree::set_machine_problem_type(EProblemType mode)
 {
-	m_mode=mode;
+	m_mode = mode;
 }
 
 bool CARTree::is_label_valid(std::shared_ptr<Labels> lab) const
 {
-	if (m_mode==PT_MULTICLASS && lab->get_label_type()==LT_MULTICLASS)
+	if (m_mode == PT_MULTICLASS && lab->get_label_type() == LT_MULTICLASS)
 		return true;
-	else if (m_mode==PT_REGRESSION && lab->get_label_type()==LT_REGRESSION)
+	else if (m_mode == PT_REGRESSION && lab->get_label_type() == LT_REGRESSION)
 		return true;
 	else
 		return false;
 }
 
-std::shared_ptr<MulticlassLabels> CARTree::apply_multiclass(std::shared_ptr<Features> data)
+std::shared_ptr<MulticlassLabels>
+CARTree::apply_multiclass(std::shared_ptr<Features> data)
 {
 	require(data, "Data required for classification in apply_multiclass");
 
 	// apply multiclass starting from root
-	auto current=get_root()->as<bnode_t>();
+	auto current = get_root()->as<bnode_t>();
 
 	require(current, "Tree machine not yet trained.");
-	return apply_from_current_node(data->as<DenseFeatures<float64_t>>(), current)->as<MulticlassLabels>();
+	if (data->get_name() == "DenseSubSamplesFeatures")
```

Comment:
You shouldn't compare strings, you can check the result of dynamic_cast. Btw SGObject has a `as` member function to do the casting:
```cpp 
if (auto* tmp = data->as<DenseSubSamplesFeatures<float64_t>>(data))
```

Reply:
Also, doesn't get_name() return `const char*`? If so what you're doing here is compare pointer values, what you should use is [strcmp](https://en.cppreference.com/w/c/string/byte/strcmp)

Reply:
Thanks for your review, I find that auto formatting make my commit a mess, so i closed this pr, i will create a new pr.

Reply:
OK! I was wondering why it was closed. I will review it better once you updated it. 
Btw you don't need to close this PR, just continue working here and let us know when it's ready for review!